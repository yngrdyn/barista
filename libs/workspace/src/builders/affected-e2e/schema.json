{
  "$schema": "http://json-schema.org/schema",
  "id": "affected-e2e",
  "type": "object",
  "properties": {
    "e2eTarget": {
      "type": "string",
      "description": "The target configuration that should be used."
    },
    "testAppSourceFolder": {
      "type": "string",
      "description": "The app root of the end to end test that should be scanned for modules using the affected components"
    }
  },
  "additionalProperties": false,
  "required": ["e2eTarget", "testAppSourceFolder"],
  "definitions": {}
}
