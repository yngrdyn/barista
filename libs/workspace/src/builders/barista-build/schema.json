{
  "$schema": "https://json-schema.org/draft-07/schema",
  "$id": "https://json-schema.org/draft-07/schema",
  "title": "Workspace builder",
  "description": "",
  "type": "object",
  "properties": {
    "outputPath": {
      "type": "string",
      "description": "The output path of the generated files."
    },
    "devServerTarget": {
      "type": "string",
      "description": "The universal serve-ssr target that should be started."
    },
    "routesFile": {
      "type": "string",
      "description": "Path to the file that holds the route information."
    },
    "routes": {
      "type": "string",
      "description": "Comma separated list of routes."
    }
  }
}
