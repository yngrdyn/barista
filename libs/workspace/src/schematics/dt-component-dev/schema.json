{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentDev",
  "title": "DtComponent Dev Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for your Dev component?"
    }
  }
}
