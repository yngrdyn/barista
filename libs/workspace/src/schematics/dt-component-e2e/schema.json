{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentE2E",
  "title": "DtComponent E2E Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for your e2e component?"
    }
  }
}
