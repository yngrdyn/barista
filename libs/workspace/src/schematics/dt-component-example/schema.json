{
  "$schema": "http://json-schema.org/schema",
  "id": "DtComponentExample",
  "title": "DtComponent Example Schema",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the example",
      "default": "default",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What Name would you like to use for the example?"
    },
    "component": {
      "type": "string",
      "description": "Name of the component",
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": "What component does the example belong to?"
    }
  }
}
