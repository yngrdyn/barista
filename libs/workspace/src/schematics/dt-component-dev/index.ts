/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { strings } from '@angular-devkit/core';
import {
  apply,
  chain,
  mergeWith,
  Rule,
  template,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { join } from 'path';
import * as ts from 'typescript';
import { findNodes, getSourceFile, getIndentation } from '../utils/ast-utils';
import { commitChanges, InsertChange } from '../utils/change';
import { generateComponentOptions } from '../utils/schematics-utils';
import { DtComponentDevOptions } from './schema';
import { extend } from 'lodash-es';

interface DtDevExtendedOptions {
  selector: string;
  componentModule: {
    name: string;
    package: string;
  };
  devComponent: {
    component: string;
  };
  name: string;
}

function generateDevComponentOptions(name: string): { component: string } {
  return {
    component: `${strings.classify(name)}Demo`,
  };
}

function addImport(
  sourceFile: ts.SourceFile,
  name: string,
  modulePath: string,
): InsertChange {
  const lastImport = findNodes(
    sourceFile,
    ts.SyntaxKind.ImportDeclaration,
  ).pop() as ts.ImportDeclaration;
  const end = lastImport.end + 1;
  const dashName = strings.dasherize(name);
  const classyName = strings.classify(name);
  const toInsertImport = `import { ${classyName}Demo } from './${dashName}/${dashName}-demo.component';
  `;
  return new InsertChange(modulePath, end, toInsertImport);
}

/**
 * Adds a new route inside the demos routes
 */
export function addRoute(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'devapp-routing.module.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const importChange = addImport(sourceFile, options.name, modulePath);

    const routesDeclaration = findNodes(
      sourceFile,
      ts.SyntaxKind.VariableDeclaration,
    ).find(
      (node: ts.VariableDeclaration) => node.name.getText() === 'routes',
    ) as ts.VariableDeclaration;

    const routesElements = (routesDeclaration.initializer as ts.ArrayLiteralExpression)
      .elements;
    const lastElement = routesElements[routesElements.length - 1];
    const endRoutes = routesElements.hasTrailingComma
      ? lastElement.getEnd() + 1
      : lastElement.getEnd();

    const toInsertRoute = `${getIndentation(
      routesElements,
    )}{ path: '${strings.dasherize(
      options.name,
    )}', component: ${strings.classify(options.name)}Demo },`;
    const routeChange = new InsertChange(modulePath, endRoutes, toInsertRoute);

    return commitChanges(host, [importChange, routeChange], modulePath);
  };
}

/**
 * Adds a new navigation item into demos.
 */
export function addNavItem(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'devapp.component.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const navItemDeclaration = findNodes(
      sourceFile,
      ts.SyntaxKind.PropertyDeclaration,
    ).find(
      (node: ts.PropertyDeclaration) => node.name.getText() === 'navItems',
    ) as ts.PropertyDeclaration;

    const navItems = (navItemDeclaration.initializer as ts.ArrayLiteralExpression)
      .elements;
    const lastNavItem = navItems[
      navItems.length - 1
    ] as ts.ObjectLiteralExpression;
    const elementsEnd = navItems.hasTrailingComma
      ? lastNavItem.getEnd() + 1
      : lastNavItem.getEnd();
    const navChange = new InsertChange(
      modulePath,
      elementsEnd,
      `${getIndentation(navItems)}{ name: '${strings.classify(
        options.name,
      )}', route: '/${strings.dasherize(options.name)}' },`,
    );

    return commitChanges(host, [navChange], modulePath);
  };
}

/**
 * Add module.
 */
export function addModule(options: DtDevExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join('apps', 'dev', 'src', 'app.module.ts');
    const sourceFile = getSourceFile(host, modulePath);

    const importChange = addImport(sourceFile, options.name, modulePath);

    const moduleDeclarations = findNodes(
      sourceFile,
      ts.SyntaxKind.PropertyAssignment,
    ).find(
      (node: ts.PropertyAssignment) => node.name.getText() === 'declarations',
    ) as ts.PropertyAssignment;

    const moduleDeclarationArray = moduleDeclarations.initializer as ts.ArrayLiteralExpression;
    const moduleElements = moduleDeclarationArray.elements;
    const lastModule = moduleElements[moduleElements.length - 1];
    const modulesEnd = moduleElements.hasTrailingComma
      ? lastModule.getEnd() + 1
      : lastModule.getEnd();
    const moduleToInsert = `${getIndentation(moduleElements)}${strings.classify(
      options.name,
    )}Demo`;

    const moduleChange = new InsertChange(
      modulePath,
      modulesEnd,
      moduleToInsert,
    );

    return commitChanges(host, [importChange, moduleChange], modulePath);
  };
}

export default function (options: DtComponentDevOptions): Rule {
  return async (_host: Tree) => {
    const extendedOptions: DtDevExtendedOptions = {
      ...options,
      selector: `dt-${strings.dasherize(options.name)}`,
      componentModule: generateComponentOptions(options.name),
      devComponent: generateDevComponentOptions(options.name),
    };

    const templateSource = apply(url('./files'), [
      template({
        ...strings,
        ...extendedOptions,
      }),
    ]);

    return chain([
      mergeWith(templateSource),
      addRoute(extendedOptions),
      addNavItem(extendedOptions),
      addModule(extendedOptions),
    ]);
  };
}
