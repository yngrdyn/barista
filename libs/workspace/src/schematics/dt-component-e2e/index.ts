/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { strings } from '@angular-devkit/core';
import {
  apply,
  chain,
  mergeWith,
  Rule,
  template,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { formatFiles } from '@nrwl/workspace';
import { join } from 'path';
import * as ts from 'typescript';
import { findNodes, getIndentation, getSourceFile } from '../utils/ast-utils';
import { commitChanges, InsertChange } from '../utils/change';
import { DtComponentE2EOptions } from './schema';
interface DtE2EExtendedOptions {
  selector: string;
  componentModule: {
    name: string;
    package: string;
  };
  e2eComponent: {
    component: string;
    module: string;
  };
  name: string;
}

function generateComponentOptions(
  name: string,
): { name: string; package: string } {
  return {
    name: `Dt${strings.classify(name)}Module`,
    package: `@dynatrace/barista-components/${strings.dasherize(name)}`,
  };
}

function generateE2EComponentOptions(
  name: string,
): { component: string; module: string } {
  return {
    component: `DtE2E${strings.classify(name)}`,
    module: `DtE2E${strings.classify(name)}Module`,
  };
}

/**
 * Adds a new route inside the ui-test routes
 */
export function addRoute(options: DtE2EExtendedOptions): Rule {
  return (host: Tree) => {
    const modulePath = join(
      'apps',
      'components-e2e',
      'src',
      'app',
      'app.routing.module.ts',
    );
    const sourceFile = getSourceFile(host, modulePath);

    /**
     * find last route and add new route
     */
    const routesDeclaration = findNodes(
      sourceFile,
      ts.SyntaxKind.VariableDeclaration,
    ).find(
      (node: ts.VariableDeclaration) => node.name.getText() === 'routes',
    ) as ts.VariableDeclaration;
    const routesElements = (routesDeclaration.initializer as ts.ArrayLiteralExpression)
      .elements;
    const lastElement = routesElements[routesElements.length - 1];
    const end = routesElements.hasTrailingComma
      ? lastElement.getEnd() + 1
      : lastElement.getEnd();
    const toInsert = `{
    path: '${strings.dasherize(options.name)}',
    loadChildren: () =>
      import('../components/${strings.dasherize(
        options.name,
      )}/${strings.dasherize(options.name)}.module').then(
        (module) => module.${options.e2eComponent.module},
      ),
  },`;
    const routesChange = new InsertChange(modulePath, end, toInsert);

    return commitChanges(host, [routesChange], modulePath);
  };
}

export default function (options: DtComponentE2EOptions): Rule {
  return async () => {
    const extendedOptions: DtE2EExtendedOptions = {
      ...options,
      selector: `dt-${strings.dasherize(options.name)}`,
      componentModule: generateComponentOptions(options.name),
      e2eComponent: generateE2EComponentOptions(options.name),
    };

    const templateSource = apply(url('./files'), [
      template({
        ...strings,
        ...extendedOptions,
      }),
    ]);

    return chain([
      mergeWith(templateSource),
      addRoute(extendedOptions),
      formatFiles(),
    ]);
  };
}
