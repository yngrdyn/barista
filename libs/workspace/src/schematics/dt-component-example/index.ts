/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { strings } from '@angular-devkit/core';
import {
  apply,
  chain,
  mergeWith,
  Rule,
  template,
  move,
  Tree,
  url,
  noop,
} from '@angular-devkit/schematics';
import { formatFiles } from '@nrwl/workspace';
import { DtComponentExampleOptions } from './schema';
import { normalize } from '@angular-devkit/core';
import { LICENSE_HEADER } from '../../utils/common-utils';
import { join } from 'path';
import {
  getSourceFile,
  findNodes,
  addImport,
  addExport,
} from '../utils/ast-utils';
import * as ts from 'typescript';
import { commitChanges, InsertChange } from '../utils/change';
import {
  updateExamplesModule,
  changeRoutingModule,
  changeNavigation,
  updateExamplesBarrel,
} from './example-tools.utils';

export interface DtExampleExtendedOptions {
  componentSelector: string;
  examplesModule: string;
  exampleRoute: string;
  exampleId: string;
  selector: string;
  componentModule: {
    name: string;
    package: string;
  };
  exampleComponent: {
    component: string;
    module: string;
    modulesConstant: string;
    template: string;
  };
  name: string;
  component: string;
  dashName: string;
}

const EXAMPLES_SOURCE_PATH = 'libs/examples/src/';

/**
 * Function to generate util options for the component level generation.
 * @param options
 */
function generateComponentOptions(
  options: DtComponentExampleOptions,
): { name: string; package: string } {
  return {
    name: `Dt${strings.classify(options.component)}Module`,
    package: `@dynatrace/barista-components/${strings.dasherize(
      options.component,
    )}`,
  };
}

/**
 * Function to generate util options to be used during example generation.
 * @param options
 */
function generateExampleComponentOptions(
  options: DtComponentExampleOptions,
): {
  component: string;
  module: string;
  modulesConstant: string;
  template: string;
} {
  const dashName = strings.dasherize(options.component);
  return {
    component: `DtExample${strings.classify(options.name)}${strings.classify(
      options.component,
    )}`,
    module: `DtExample${strings.classify(options.name)}Module`,
    modulesConstant: `DT_${strings
      .underscore(options.component)
      .toUpperCase()}_EXAMPLES`,
    template: `${dashName}-${strings.dasherize(options.name)}-example.html`,
  };
}

/**
 * Generate the index file for the examples.
 * @param options
 */
function generateBarrelFileContent(options: DtExampleExtendedOptions): string {
  return `${LICENSE_HEADER}

  export * from './${options.dashName}-examples.module';
  export * from './${options.exampleRoute}';
  `;
}

/**
 * Generate the module file for the examples.
 * @param options
 */
function generateModuleFileContent(options: DtExampleExtendedOptions): string {
  return `${LICENSE_HEADER}
  import { NgModule } from '@angular/core';
  import { ${options.componentModule.name} } from '${options.componentModule.package}';
  import { ${options.exampleComponent.component} } from './${options.exampleRoute}';

  @NgModule({
    imports: [${options.componentModule.name}],
    declarations: [${options.exampleComponent.component}],
  })
  export class ${options.examplesModule} {}`;
}

/**
 * Function to update the module files.
 * @param options
 */
function updateModules(options: DtExampleExtendedOptions): Rule {
  return (host: Tree) => {
    try {
      // add import
      const modulePath = join(
        'libs',
        'examples',
        'src',
        options.dashName,
        `${options.dashName}-examples.module.ts`,
      );
      const sourceFile = getSourceFile(host, modulePath);

      const importChange = addImport(
        sourceFile,
        options.exampleComponent.component,
        options.exampleRoute,
        modulePath,
      );
      // find last module in the array, and add new module
      const modulesDeclaration = findNodes(
        sourceFile,
        ts.SyntaxKind.PropertyAssignment,
      ).find(
        (node: ts.PropertyAssignment) => node.name.getText() === 'declarations',
      ) as ts.PropertyAssignment;
      const modulesElements = (modulesDeclaration.initializer as ts.ArrayLiteralExpression)
        .elements;
      const lastElement = modulesElements[modulesElements.length - 1];
      const end = modulesElements.hasTrailingComma
        ? lastElement.getEnd() + 1
        : lastElement.getEnd();
      const routesChange = new InsertChange(
        modulePath,
        end,
        `${modulesElements.hasTrailingComma ? ' ' : ', '}${
          options.exampleComponent.component
        }`,
      );

      return commitChanges(host, [importChange, routesChange], modulePath);
    } catch (error) {
      throw new Error(
        `There was a problem updating ${options.dashName}-examples.module.ts ${error.message}`,
      );
    }
  };
}

/**
 * Function to update the index file.
 * @param options
 */
function updateIndex(options: DtExampleExtendedOptions): Rule {
  return (host: Tree) => {
    try {
      // add export to index.ts
      const indexPath = join(
        'libs',
        'examples',
        'src',
        options.dashName,
        'index.ts',
      );
      const sourceIndexFile = getSourceFile(host, indexPath);
      const exportChange = addExport(
        sourceIndexFile,
        options.exampleRoute,
        indexPath,
      );

      return commitChanges(host, exportChange, indexPath);
    } catch (error) {
      throw new Error(
        `There was a problem updating 'libs/examples/src/${options.dashName}/index.ts' : ${error.message}`,
      );
    }
  };
}

/**
 * The main function of the example schematics.
 * @param options
 */
export default function (options: DtComponentExampleOptions): Rule {
  return async (tree: Tree) => {
    const dashName = strings.dasherize(options.component);
    const exampleId = `${dashName}-${strings.dasherize(options.name)}`;
    const extendedOptions: DtExampleExtendedOptions = {
      ...options,
      componentSelector: `dt-${dashName}`,
      selector: `dt-example-${strings.dasherize(options.name)}-${dashName}`,
      componentModule: generateComponentOptions(options),
      exampleComponent: generateExampleComponentOptions(options),
      examplesModule: `DtExamples${strings.classify(options.component)}Module`,
      exampleRoute: `${exampleId}-example/${exampleId}-example`,
      exampleId,
      dashName: dashName,
    };

    // Generating files from file templates, and moving them to their path.
    const templateSource = apply(url('./files'), [
      template({
        ...strings,
        ...extendedOptions,
      }),
      move(
        normalize(
          `${EXAMPLES_SOURCE_PATH}${dashName}/${dashName}-${strings.dasherize(
            options.name,
          )}-example`,
        ),
      ),
    ]);

    const indexPath = normalize(`${EXAMPLES_SOURCE_PATH}${dashName}/index.ts`);

    const isNewComponent = !tree.exists(indexPath);

    // If the example is for a new component, so there are no other examples for
    // for this component yet, we have to generate the index and module files.
    if (isNewComponent) {
      tree.create(indexPath, generateBarrelFileContent(extendedOptions));

      const modulePath = normalize(
        `${EXAMPLES_SOURCE_PATH}${dashName}/${dashName}-examples.module.ts`,
      );
      tree.create(modulePath, generateModuleFileContent(extendedOptions));
    }

    return chain([
      mergeWith(templateSource),
      isNewComponent ? noop() : updateModules(extendedOptions),
      isNewComponent ? noop() : updateIndex(extendedOptions),
      isNewComponent ? updateExamplesModule(extendedOptions) : noop(),
      updateExamplesBarrel(extendedOptions, isNewComponent),
      changeRoutingModule(extendedOptions),
      changeNavigation(extendedOptions, isNewComponent),
      formatFiles(),
    ]);
  };
}
