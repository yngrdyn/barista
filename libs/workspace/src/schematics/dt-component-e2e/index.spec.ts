/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import { promises as fs } from 'fs';
import { join } from 'path';

export async function getFixture(filePath: string): Promise<string> {
  const fixturesFolder = join(__dirname, './fixtures');
  return fs.readFile(join(fixturesFolder, filePath), {
    encoding: 'utf-8',
  });
}

export async function addFixtureToTree(
  tree: Tree,
  source: string,
  destination: string,
): Promise<void> {
  const content = await getFixture(source);
  if (tree.exists(destination)) {
    tree.overwrite(destination, content);
    return;
  }
  tree.create(destination, content);
}

const getNewFiles = (newFiles: string[], oldFiles: string[]): string[] =>
  newFiles.filter((file) => !oldFiles.includes(file));

describe('Add E2E Component', () => {
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    newProjectRoot: 'apps',
    version: '8.0.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'myapp',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: true,
    skipPackageJson: true,
  };

  let initialTree: UnitTestTree;
  let schematicRunner: SchematicTestRunner;
  let initialFiles: string[];

  beforeEach(async () => {
    schematicRunner = new SchematicTestRunner(
      '@dynatrace/barista-components/schematics',
      join(__dirname, '../../../collection.json'),
    );

    initialTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions,
        Tree.empty(),
      )
      .toPromise();

    initialTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        initialTree,
      )
      .toPromise();

    await addFixtureToTree(
      initialTree,
      'app.routing.module.ts.fixture.ts',
      'apps/components-e2e/src/app/app.routing.module.ts',
    );

    initialFiles = initialTree.files;
  });

  it('should add the necessary files', async () => {
    const result = await schematicRunner
      .runSchematicAsync('dt-e2e', { name: 'alert' }, initialTree)
      .toPromise();

    const newFiles = getNewFiles(result.files, initialFiles);

    expect(newFiles).toEqual([
      '/apps/components-e2e/src/components/alert/alert.e2e.ts',
      '/apps/components-e2e/src/components/alert/alert.html',
      '/apps/components-e2e/src/components/alert/alert.module.ts',
      '/apps/components-e2e/src/components/alert/alert.po.ts',
      '/apps/components-e2e/src/components/alert/alert.ts',
    ]);
  });

  it('should generate files with the appropriate content', async () => {
    const result = await schematicRunner
      .runSchematicAsync('dt-e2e', { name: 'alert' }, initialTree)
      .toPromise();

    const newFiles = getNewFiles(result.files, initialFiles);
    newFiles.map((file) => expect(result.readContent(file)).toMatchSnapshot());
  });

  it('should change files appropriately', async () => {
    const result = await schematicRunner
      .runSchematicAsync('dt-e2e', { name: 'alert' }, initialTree)
      .toPromise();

    expect(
      result.readContent('apps/components-e2e/src/app/app.routing.module.ts'),
    ).toMatchSnapshot();
  });

  it('should throw error if we already have the files in there', async () => {
    await addFixtureToTree(
      initialTree,
      'alert.fixture.ts',
      '/apps/components-e2e/src/components/alert/alert.ts',
    );

    await expect(
      schematicRunner
        .runSchematicAsync('dt-e2e', { name: 'alert' }, initialTree)
        .toPromise(),
    ).rejects.toThrow();
  });
});
