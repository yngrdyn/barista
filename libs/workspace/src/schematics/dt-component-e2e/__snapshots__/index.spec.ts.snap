// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add E2E Component should change files appropriately 1`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// @ts-nocheck
// Code copy pasted from original app.routing.module
// Used only for testing purposes
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

export const routes: Routes = [
  {
    path: 'autocomplete',
    loadChildren: () =>
      import('').then((module) => module.DtE2EAutocompleteModule),
  },
  {
    path: 'button',
    loadChildren: () => import('').then((module) => module.DtE2EButtonModule),
  },
  {
    path: 'alert',
    loadChildren: () =>
      import('../components/alert/alert.module').then(
        (module) => module.DtE2EAlertModule,
      ),
  },
];

@NgModule({
  imports: [
    CommonModule,
    RouterModule.forRoot(routes, {
      scrollPositionRestoration: 'enabled',
      anchorScrolling: 'enabled',
      paramsInheritanceStrategy: 'always',
      enableTracing: false, // Can be set for debugging the router
      initialNavigation: 'enabled',
    }),
  ],
  exports: [RouterModule],
})
export class AppRoutingModule {}
"
`;

exports[`Add E2E Component should generate files with the appropriate content 1`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { alert } from './alert.po';

fixture('Alert').page('http://localhost:4200/alert');

test('should exists', async (testController: TestController) => {
  await testController.expect(alert.exists).ok();
});
"
`;

exports[`Add E2E Component should generate files with the appropriate content 2`] = `
"<dt-alert id=\\"test-alert\\"></dt-alert>
"
`;

exports[`Add E2E Component should generate files with the appropriate content 3`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { Route, RouterModule } from '@angular/router';
import { DtAlertModule } from '@dynatrace/barista-components/alert';
import { DtE2EAlert } from './alert';

const routes: Route[] = [{ path: '', component: DtE2EAlert }];

@NgModule({
  declarations: [DtE2EAlert],
  imports: [CommonModule, RouterModule.forChild(routes), DtAlertModule],
})
export class DtE2EAlertModule {}
"
`;

exports[`Add E2E Component should generate files with the appropriate content 4`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Selector } from 'testcafe';

export const alert = Selector('#test-alert');
"
`;

exports[`Add E2E Component should generate files with the appropriate content 5`] = `
"/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@angular/core';

@Component({
  selector: 'dt-e2e-alert',
  templateUrl: 'alert.html',
})
export class DtE2EAlert {}
"
`;
